File: backend/src/dispatcher/entry-queue.ts
import Heap from "heap-js";
import { Task, TaskStatus } from "./task";
import { NOOP } from "./utils/noop";

export const MAX_PRIORITY = Number.MAX_SAFE_INTEGER;

/** Priority queue with filter by min_price */
export class EntryQueue {

  // Priority might be a timestamp + delta depends on priority, so old tasks are always first
  private queue = new Heap((a: Task, b: Task) => a.getPriority() - b.getPriority());

  addTask(task: Task, priority: number): void {
    task.setPriority(priority);
    this.queue.add(task);

    task.setStatus(TaskStatus.PushedIntoQueue);
    this.taskAdded(task);
  }

  removeTask(task: Task): void {
    this.queue.remove(task);
  }

  /** Returns the first task from the priority queue with the max_price > min_price */
  popTask(min_price?: number): Task | undefined {
    for (const task of this.queue) {
      if (min_price === undefined || task.getMaxPrice() > min_price) {
        this.queue.remove(task);
        return task;
      }
    }
    return undefined;
  }

  taskAdded: (task: Task) => void = NOOP;
}

File: backend/src/dispatcher/entry_queue.py
from heapq import Heap
from task import Task, TaskStatus
from dispatcher.utils.noop import NOOP
from typing import Callable

MAX_PRIORITY = 2**53 - 1  # Number.MAX_SAFE_INTEGER

class EntryQueue:
    def init(self):
        self.queue = Heap(lambda a, b: a.getPriority() - b.getPriority())
    
    def addTask(self, task: Task, priority: int) -> None:
        task.setPriority(priority)
        self.queue.add(task)
        
        task.setStatus(TaskStatus.PushedIntoQueue)
        self.taskAdded(task)
    
    def removeTask(self, task: Task) -> None:
        self.queue.remove(task)
    
    def popTask(self, min_price: int = None) -> Task or None:
        for task in self.queue:
            if min_price is None or task.getMaxPrice() > min_price:
                self.queue.remove(task)
                return task
        return None

    taskAdded: Callable[[Task], None] = NOOP
